### ISSUE TITLE
Implement static checks for Scrapy projects

### ISSUE THREAD

**Issue by Gallaecio** [scrapy/scrapy#4421]
On static analysis specify to Scrapy projects, at the moment there only seems to be https://github.com/stummjr/flake8-scrapy

I think we should come up with additional checks based on our day-to-day experiences using Scrapy and common issues we see in the community, specially in new users, and implement them in https://github.com/stummjr/flake8-scrapy (or a fork if needed) when possible, and look for alternatives in case of checks that cannot be implemented that way.

**Comment by Gallaecio** [comment]
https://github.com/scrapy/scrapy/issues/4302, about typos in setting names, is one issue that could be addressed, but it may be better [addressed at run time](https://github.com/scrapy/scrapy/issues/4302#issuecomment-596271546). 

@Prime-5 has been working on type checking for setting values.

I’m thinking https://github.com/scrapy/scrapy/pull/4170 may be better done with a static checked, to allow for a slightly cleaner Scrapy code base.

**Comment by Prime-5** [comment]
@Gallaecio @ejulio @elacuesta I've written a draft proposal for this idea, could you give me some feedback? 
https://docs.google.com/document/d/1su4A8bzSTKgUulfHj55cLiuRqFbWOL8VbRt16kGCSPs/edit?usp=sharing
The comment rights to this doc is public.

**Comment by Gallaecio** [comment]
It may be better to decide which checks to implement now, as part of the proposal, instead of planning to do that during the summer. We don’t want to start the project to find out we don’t know what we want to check.

https://www.reddit.com/r/scrapy/ may be a good place to ask the active members of the community for feedback about the kind of static checks that they would like to see implemented.

You may also want to try and contact the author or https://github.com/stummjr/flake8-scrapy. Has the author abandoned the project? Otherwise, will they be willing accept pull requests for additional checks? Do they have any other check in mind, which they have not yet implemented due to a lack of time?

**Comment by Gallaecio** [comment]
Pylint reports `no-member` on signal assignation, such as the first example at https://docs.scrapy.org/en/latest/topics/signals.html#signals, because the target method does not exist in `Spider`.

However, it is defined in the `Spider` subclass, and things seem to work. So it would be great to have a way to improve the handling of that check for Scrapy projects.

**Comment by Prime-5** [comment]
@Gallaecio, thanks for the issue. As suggested I had asked on reddit (and quora as well) on the same day, and attempted to contact stummjr as well, but none have been fruitful yet. 

As for the ```no-member``` issue, I've reproduced it, and my thinking is that if we make a customized version of Pylint for scrapy in which we can make Pylint consider the functions defined in the subclasses (along with the ones defined in the original class), then this kind of error would be done away with.

I don't think a pylint-scrapy repo has been made yet, has it?

**Comment by Gallaecio** [comment]
> I don't think a pylint-scrapy repo has been made yet, has it?

I don’t think so either.

**Comment by Prime-5** [comment]
@Gallaecio, I have updated the draft proposal with a tentative list of checks that I'm planning to implement - [GSOC'20 Draft Proposal: Static Analyser for Scrapy](https://docs.google.com/document/d/1su4A8bzSTKgUulfHj55cLiuRqFbWOL8VbRt16kGCSPs/edit?usp=sharing), could you have a look at it once, as I'm planning to submit my proposal soon.

Also, I've kept a little time away for any more tests that get added to the list along the way.

**Comment by guptamanav077** [comment]
Hello,

I wanted to know how many candidates are you considering taking for this specific project?

Manav

**Comment by Gallaecio** [comment]
For each idea there can only be 1 student selected. For Scrapy as a project, or for Scrapinghub-powered open source projects in general, the number of student proposals that we can accept is out of our hands, it depends on how many slots Google can assign to the Python Software Foundation, and how many of those the PSF can assign to us.

**Comment by Gallaecio** [comment]
For future reference: https://www.reddit.com/r/scrapy/comments/fjok5q/ideas_for_implementing_static_checks_for_scrapy/

**Comment by Gallaecio** [comment]
Ideas:
* `SelectorList.extract` → `SelectorList.getall`
* `SelectorList.extract_first` → `SelectorList.get`
* `Selector.extract` → `Selector.get`
* `SelectorList.getall()[0]` → `SelectorList.get()`

**Comment by joaquingx** [comment]
Hey :smile:, I think it would be a valuable addition to scrapy but would be necessary to improve visibility/usage of [flake8-scrapy](https://github.com/stummjr/flake8-scrapy) right?, after apply some of these recommendations it would be added to [requirements](https://github.com/scrapy/scrapy/blob/195f738bbaa7c090df795161390056c3c6c7f1f7/setup.py#L21)?

**Comment by eLRuLL** [comment]
@joaquingx I don't think this should be added as a scrapy requirement, as "linting" is not really something necessary for scrapy to work correctly.

Now, talking about giving visibility of the project, maybe we could find a way to include this in the documentation, as a recommendation for developers

**Comment by Aman-Godara** [comment]
Hey @Gallaecio,
I am looking forward to take part in GSoC 2021. While I was going through the proposed projects by organizations I reached here. 
Among the many that I searched for, this one looked best to work on to me. I wanted to know more about the project [Static Analysis Tooling](https://gsoc2021.zyte.com/ideas). Can you please tell where can I start reading about this?

**Comment by Gallaecio** [comment]
Another idea from https://github.com/scrapy/scrapy/pull/6851: detect the use of static start URLs, and recommend setting `allowed_domains` accordingly.

**Comment by Gallaecio** [comment]
An idea, although a bit specific to scrapy-poet: if scrapy-poet is installed, inspect values passed to callback params of Request and make sure that none of the typings of the callback signature are under `if TYPE_CHECKING`.

**Comment by Gallaecio** [comment]
On top of some of the changes to the default Scrapy behavior being proposed in #6597, I wonder if there are some things that it would make sense to implement in a static checker.

I am specifically thinking of the `USER_AGENT` setting, for example. It would be nice if a static checker for Scrapy would flag as an issue both not defining that setting and defining it with a “sneaky” value (e.g. an user-agent string mimicking a browser).

It may also make sense to flag increases of concurrency-related settings in `settings.py`, recommending instead for users to be considerate, learn about the estimated traffic of their target websites with tools like https://www.similarweb.com/, and adjust concurrency settings accordingly, and set them either per-spider or through the `DOWNLOAD_SLOTS` setting.

In other words, use static checks to do our best  for users not to accidentally do the wrong thing, and have them be explicit when they go against best practices by having to disable rules named e.g. `polite-user-agent` or `polite-concurrency-per-domain`.

**Comment by Gallaecio** [comment]
It would be nice to have static checks for deprecated APIs (e.g. `start_requests()`), specially if we can implement opt-in automatic fixes for them.

**Comment by wRAR** [comment]
> It would be nice to have static checks for deprecated APIs (e.g. start_requests()),

And `extract_first()`

**Comment by Nandith-0777** [comment]
I went through the discussion and the links in this issue, including flake8-scrapy and the later ideas mentioned here.

From what I can see, this has grown into a broader question of what static analysis for Scrapy should look like today, especially given newer tools like ruff and scrapy-lint.

As a starting point, would it make sense to focus on one small, concrete check based on common Scrapy pitfalls (for example deprecated selector methods like extract() / extract_first(), or meta=response.meta usage), and implement it in whichever toolchain is preferred now?

I’d appreciate guidance on which tool (flake8, pylint, ruff, etc.) and which single check would be a good first step.
